### PR TITLE
Add a couple of admin monitors for executors on the controller

### PR DESCRIPTION
**Pull request by jeffret-b**
Running jobs on executors on the controller can have some negative consequences. They can compete for resources. As noted in the old documentation there can be [security implications](https://wiki.jenkins.io/display/JENKINS/Security+implication+of+building+on+master). (This documentations needs to be migrated and updated but that's separate from this current PR.) Unfortunately, administrators often don't know about these concerns and leave executors configured on the controller.

This PR adds two new admin monitors to warn admins. One is activated when there are no agents or clouds defined, recommending they be set up. The other is activated after an agent or a cloud has been defined. The two are separate to help guide the admin to a proper setup through the different steps.

<details><summary>Screenshots</summary>

Screen Shot 2021-03-05 at 12 31 15 PM

Screen Shot 2021-03-05 at 12 31 54 PM

</details>

### Proposed changelog entries

* Entry 1: Add administrative monitors recommending no executors are configured on the controller.

### Proposed upgrade guidelines

N/A

Should not need any upgrade guidelines. As with all other admin monitors, these can be dismissed if the admin wishes to ignore them.

### Submitter checklist

- [ ] (If applicable) Jira issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [x] Appropriate autotests or explanation to why this change has no tests
- [n/a] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

File: core/src/main/java/jenkins/diagnostics/ControllerExecutorsAgents.java
Comment (daniel-beck) [inline core/src/main/java/jenkins/diagnostics/ControllerExecutorsAgents.java:33]:
```suggestion
@Symbol("controllerExecutorsWithAgents")
```

File: core/src/main/java/jenkins/diagnostics/ControllerExecutorsNoAgents.java
Comment (daniel-beck) [inline core/src/main/java/jenkins/diagnostics/ControllerExecutorsNoAgents.java:33]:
```suggestion
@Symbol("controllerExecutorsWithoutAgents")
```

File: core/src/main/resources/jenkins/diagnostics/ControllerExecutorsAgents/message.properties
Comment (daniel-beck) [inline core/src/main/resources/jenkins/diagnostics/ControllerExecutorsAgents/message.properties:23]:
```suggestion
ExecutorsWarning=Distributed builds are set up so consider disabling builds on the controller. See <a href="https://www.jenkins.io/redirect/building-on-master/" target="_blank" rel="noopener noreferrer">Security implications of building on master</a>.
```

Needs corresponding PR to jenkins.io (can be done at the same time as merging this PR) to introduce the redirect.

See also https://github.com/jenkinsci/jenkins/pull/4910 for when I introduced these attributes.

File: core/src/main/resources/jenkins/diagnostics/ControllerExecutorsNoAgents/message.properties
Comment (daniel-beck) [inline core/src/main/resources/jenkins/diagnostics/ControllerExecutorsNoAgents/message.properties:23]:
```suggestion
ExecutorsWarning=You should set up distributed builds. Building on the controller node can be a security issue. See <a href="https://www.jenkins.io/redirect/building-on-master/" target="_blank" rel="noopener noreferrer">Security implications of building on master</a>.
```

**Review by daniel-beck — CHANGES_REQUESTED** [review]
Only minor issues 😃

**Comment by daniel-beck** [comment]
> This PR adds two new admin monitors to warn admins. One is activated when there are no agents or clouds defined, recommending they be set up. The other is activated after an agent or a cloud has been defined. The two are separate to help guide the admin to a proper setup through the different steps.

One big advantage of two monitors is also that the second monitor will fire even if the first one was dismissed right at the beginning of an instance being set up, when the admin has no notable security requirements and just wanted to get _something_ working. The instance grew, then the admin adds the first agent, and the second admin monitor will pop up, reminding them about this. Now, with an agent set up, it'll be less of a hurdle to continue with more and move all builds off the controller.

File: core/src/main/resources/jenkins/diagnostics/ControllerExecutorsNoAgents/message.jelly
Comment (daniel-beck) [inline core/src/main/resources/jenkins/diagnostics/ControllerExecutorsNoAgents/message.jelly:28]: This form target makes no sense:

* "Manage" should take me somewhere where I can configure agents or clouds. Neither is the case with this target. I expect we'll need two buttons, "Set up Agent" and "Set up Cloud" linking to `/computer/new` and `/configureClouds/` respectively (or just one link to `/computer/` but that seems less user friendly).
* With this form target, the dismiss button doesn't work at all.

There's a reason all of these monitors use a web method that barely does anything in almost all cases.

File: core/src/main/resources/jenkins/diagnostics/ControllerExecutorsAgents/message.jelly
Comment (daniel-beck) [inline core/src/main/resources/jenkins/diagnostics/ControllerExecutorsAgents/message.jelly:30]: Same form target problem as with the other monitor for 'Dismiss', but this one at least has a useful 'Manage' button.

**Review by daniel-beck — CHANGES_REQUESTED** [review]

**Comment by jeffret-b** [comment]
Maybe we should follow more the model of `QueueItemAuthenticatorMonitor`, which has a "Learn More ..." button and a "Dismiss" button. We can hook the URL to the documentation into the "Learn More ..." as that monitor does and not worry about implementing "Manage" button(s).

What do you think?

(As part of that, or whatever way we decide to go, I can get the "Dismiss" button working.)

**Comment by jeffret-b** [comment]
Updated UI / presentation as suggested.

File: core/src/main/java/jenkins/diagnostics/ControllerExecutorsAgents.java
Comment (daniel-beck) [inline core/src/main/java/jenkins/diagnostics/ControllerExecutorsAgents.java:59]: Permission checks are unnecessary, handled by `AdministrativeMonitor` being a `StaplerProxy`.

File: core/src/main/resources/jenkins/diagnostics/ControllerExecutorsNoAgents/message.jelly
Comment (daniel-beck) [inline core/src/main/resources/jenkins/diagnostics/ControllerExecutorsNoAgents/message.jelly:30]:
```suggestion
      <f:submit name="agent" value="${%Set up agent}"/>
      <f:submit name="cloud" value="${%Set up cloud}"/>
```

File: core/src/main/java/jenkins/diagnostics/ControllerExecutorsNoAgents.java
Comment (daniel-beck) [inline core/src/main/java/jenkins/diagnostics/ControllerExecutorsNoAgents.java:53]: Permission checks are unnecessary, handled by `AdministrativeMonitor` being a `StaplerProxy`.

**Review by daniel-beck — COMMENTED** [review]
Looks good. I expect to approve once I had a chance to test this.

File: core/src/main/resources/jenkins/diagnostics/ControllerExecutorsAgents/message.properties
Comment (daniel-beck) [inline core/src/main/resources/jenkins/diagnostics/ControllerExecutorsAgents/message.properties:23]: I am not sure this is the best phrasing we can use, but I don't have suggestions for alternatives.

Strictly speaking, this could just use the same message as the other monitor, emphasizing the potential security problems.

**Review by daniel-beck — APPROVED** [review]
Tested this and it works fine. Approval conditional on applying at least the suggested i18n fix.

File: core/src/main/resources/jenkins/diagnostics/ControllerExecutorsAgents/message.properties
Comment (jeffret-b) [reply core/src/main/resources/jenkins/diagnostics/ControllerExecutorsAgents/message.properties:23]: I really don't have an opinion on the text. I'm not too fond of the current wording, but don't have much better suggestions at the moment. I have a slight negative opinion on making the text relatively the same between the two.

**Comment by jeffret-b** [comment]
Updated for the latest review comments. I'm open to better suggestions on the text.

File: core/src/main/resources/jenkins/diagnostics/ControllerExecutorsAgents/message.properties
Comment (daniel-beck) [reply core/src/main/resources/jenkins/diagnostics/ControllerExecutorsAgents/message.properties:23]: > I have a slight negative opinion on making the text relatively the same between the two.

I mean, it's only separate for internal reasons (to reappear when an instance gets reconfigured to use distributed builds), so in a way it makes sense to have the same message, just slightly different buttons for the next action to take.

IMO reusing the same message would be strictly bitter than this one.

**Review by daniel-beck — APPROVED** [review]

File: core/src/main/java/jenkins/diagnostics/ControllerExecutorsAgents.java
Comment (daniel-beck) [inline core/src/main/java/jenkins/diagnostics/ControllerExecutorsAgents.java:39]: `@Restricted(NoExternalUse.class)`

File: core/src/main/java/jenkins/diagnostics/ControllerExecutorsNoAgents.java
Comment (daniel-beck) [inline core/src/main/java/jenkins/diagnostics/ControllerExecutorsNoAgents.java:39]: `@Restricted(NoExternalUse.class)`

**Review by daniel-beck — CHANGES_REQUESTED** [review]

**Review by daniel-beck — APPROVED** [review]

**Comment by jeffret-b** [comment]
What are you thinking of for the redirect, @daniel-beck ?

**Comment by daniel-beck** [comment]
> What are you thinking of for the redirect, @daniel-beck ?

Unsure what you mean, but if you're asking where the redirect should point, I think its' good enough to point to the wiki URL you originally used for now.

**Comment by jeffret-b** [comment]
That's basically what I meant, but I thought you might have had other ideas about what the redirect should look like.

**Comment by jeffret-b** [comment]
Created redirect.

**Comment by daniel-beck** [comment]
Checkstyle wasting time:

`[ERROR] /home/jenkins/workspace/Core_jenkins_PR-5337/core/src/main/resources/jenkins/diagnostics/Messages.properties:1: File does not end with a newline. [NewlineAtEndOfFile]`

**Review by res0nance — APPROVED** [review]

File: core/src/main/java/jenkins/diagnostics/ControllerExecutorsAgents.java
Comment (oleg-nenashev) [inline core/src/main/java/jenkins/diagnostics/ControllerExecutorsAgents.java:57]: A new `Nodes#hasNodes()` method could be used for a minor optimization.
Also, caching Jenkins instance in a local variable would help.

File: core/src/main/resources/jenkins/diagnostics/ControllerExecutorsAgents/message.properties
Comment (oleg-nenashev) [reply core/src/main/resources/jenkins/diagnostics/ControllerExecutorsAgents/message.properties:23]: This text is indeed confusing. The monitor does not analyze builds at all.
I would recommend writing "The controller allows running jobs on it while there are available nodes or clouds. It might be a security risk. Consider disabling executors on the controller instance"

Comment (oleg-nenashev) [inline core/src/main/resources/jenkins/diagnostics/ControllerExecutorsAgents/message.properties:23]:
```suggestion
ExecutorsWarning=The controller allows running jobs on it while there are available nodes or clouds. It might be a security risk. Consider disabling executors on the controller instance. See <a href="https://www.jenkins.io/redirect/building-on-controller/" target="_blank" rel="noopener noreferrer">the documentation</a>.
```

File: core/src/main/resources/jenkins/diagnostics/ControllerExecutorsNoAgents/message.properties
Comment (oleg-nenashev) [inline core/src/main/resources/jenkins/diagnostics/ControllerExecutorsNoAgents/message.properties:23]:
```suggestion
ExecutorsWarning=There are no configured agents or clouds in the system. Building on the controller node can be a security issue. See <a href="https://www.jenkins.io/redirect/building-on-controller/" target="_blank" rel="noopener noreferrer">the documentation</a>.
```

File: core/src/main/java/jenkins/diagnostics/ControllerExecutorsNoAgents.java
Comment (oleg-nenashev) [inline core/src/main/java/jenkins/diagnostics/ControllerExecutorsNoAgents.java:68]: same as above

**Review by oleg-nenashev — COMMENTED** [review]
The code looks good overall, but the wording could be improved

File: core/src/main/java/jenkins/diagnostics/ControllerExecutorsAgents.java
Comment (jeffret-b) [reply core/src/main/java/jenkins/diagnostics/ControllerExecutorsAgents.java:57]: We could add a method like that, but I'm not sure it would accomplish much. The class naming and usage for nodes in that area is already quite a mess and I don't if adding a utility method to `Nodes` would help much. We could add a utility to `Jenkins` to check for nodes or clouds, but it's a pretty specific, easy check.

I'm not sure that caching the Jenkins instance would help, but I can do that if desired.

**Review by jeffret-b — COMMENTED** [review]
I find the newly suggested messages more confusing than what's already there. @daniel-beck , what do you think? Anyone else have good wording suggestions.

File: core/src/main/resources/jenkins/diagnostics/ControllerExecutorsAgents/message.properties
Comment (daniel-beck) [reply core/src/main/resources/jenkins/diagnostics/ControllerExecutorsAgents/message.properties:23]: Seems redundant to explain something that the admin is very likely already aware of. I still think reusing the same message from before makes sense here.

(It's basically one monitor, but we want it to appear again once an admin set up their first agent if dismissed before, and this is how we accomplish that.)

File: core/src/main/resources/jenkins/diagnostics/ControllerExecutorsNoAgents/message.properties
Comment (daniel-beck) [reply core/src/main/resources/jenkins/diagnostics/ControllerExecutorsNoAgents/message.properties:23]: Seems unnecessary and prevents reuse (the existing message is at least somewhat ambiguous). Either works if we keep separate messages.

**Review by daniel-beck — APPROVED** [review]

File: core/src/main/resources/jenkins/diagnostics/ControllerExecutorsAgents/message.properties
Comment (jeffret-b) [inline core/src/main/resources/jenkins/diagnostics/ControllerExecutorsAgents/message.properties:23]:
```suggestion
ExecutorsWarning=You should set up distributed builds. Building on the controller node can be a security issue. See <a href="https://www.jenkins.io/redirect/building-on-controller/" target="_blank" rel="noopener noreferrer">the documentation</a>.
```

**Comment by jeffret-b** [comment]
I went with the proposal from @daniel-beck to use the same message in both cases. It works reasonably well.

We've now got a couple of approvals and I don't see any significant concerns.

**Review by oleg-nenashev — APPROVED** [review]
We may merge it in 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process

**Comment by saper** [comment]
So I think this has landed in 2.286? I have upgraded my 2.285 test instance that has 1 worker plus a working Kubernetes cloud connection configured and I got that first warning, that is

"You should set up distributed builds. Building on the controller node can be a security issue. See <a href="https://www.jenkins.io/redirect/building-on-controller/" target="_blank" rel="noopener noreferrer">the documentation</a>"

**Comment by oleg-nenashev** [comment]
@saper Yes, it did. https://www.jenkins.io/changelog/#v2.286

**Comment by daniel-beck** [comment]
@saper Please report a bug in Jira and execute each of the statements https://github.com/jenkinsci/jenkins/pull/5337/files#diff-d1779b2b9180c5f67154c02dcbcd3da3bc6968a61901ab948c691f12b40c2a4eR67-R70 individually in the script console and provide their output.

**Comment by saper** [comment]
Thank you for your very quick response - I hope https://issues.jenkins.io/browse/JENKINS-65291 is  good enough...

File: core/src/main/resources/jenkins/diagnostics/ControllerExecutorsNoAgents/message.properties
Comment (saper) [reply core/src/main/resources/jenkins/diagnostics/ControllerExecutorsNoAgents/message.properties:23]: "There are no configured agents or clouds"... is better.

File: core/src/main/resources/jenkins/diagnostics/ControllerExecutorsAgents/message.properties
Comment (saper) [inline core/src/main/resources/jenkins/diagnostics/ControllerExecutorsAgents/message.properties:23]: As per https://issues.jenkins.io/browse/JENKINS-65291 can we say something like this? "You should disable builds on the master node. You can do this be setting the number of executors to zero."

Also rename the warning to "ControllerExecutorsWarning" or something like this.

**Review by saper — COMMENTED** [review]
Personally I find the naming "ControllerExecutorsAgents" and "ControllerExecutorsNoAgents" pretty confusing. It is almost the same but not at all the same.

Also code duplication is pretty big, would that make sense to have a common base class for this? Ah, Java :)

**Pull request merged**
